File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int pv = p->val, qv = q->val;
        
        while(root != NULL){
            int rv = root->val;
            
            //if both p and q are in different subtrees, then root is our LCA
            if((pv > rv and qv < rv) || (pv < rv and qv > rv)){
                return root;
            }
            
            //if anyone becomes equal to root then its LCA by default
            else if(root == p || root == q){
                return root==p?p:q;
            }

                if(pv > rv) root = root->right;
                else root = root->left;
        }
        return NULL;
    }
};

File: 235-lowest-common-ancestor-of-a-binary-search-tree/235-lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        def lcaBST(root):
            if root == p or root == q or not root:
                return root
            left = lcaBST(root.left)
            right = lcaBST(root.right)
            
            if left and right: return root
            
            return left if left else right
        
        return lcaBST(root)